### PR TITLE
Module2 1

### PR DESCRIPTION
**Pull request by streamdp**

### CODE REVIEW

File: src/test/java/com/epam/streamdp/four/actions/GenerateStudentsTest.java [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+import java.util.logging.Level;
+import java.util.stream.Collectors;
+
+import static org.testng.Assert.assertEquals;
+
+public class GenerateStudentsTest {
+    private static final java.util.logging.Logger logger = java.util.logging.Logger.getLogger(GenerateStudentsTest.class.getName());
+    List<Student> students;
+
+    {
```
Comment (yury-barsukou) [inline src/test/java/com/epam/streamdp/four/actions/GenerateStudentsTest.java]: The code snippet could be inside a @Before* method

File: src/test/java/com/epam/streamdp/four/actions/ActionsTest.java [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+
+    @Test
+    public void testCalculateGPAForSpecificSubjectGroupAndFaculty() throws AcademicSubjectFieldCannotBeEmptyException, TheFacultyFieldMustBeSpecifiedException, TheGroupFieldMustBeSpecifiedException {
+        assertTrue(statementOfGrades.stream()
+                .filter(subject -> subject.getAcademicSubject() == AcademicSubjects.ENGLISH)
+                .filter(subject -> subject.getGroup() == 2)
+                .filter(subject -> subject.getStudent().getFaculty() == Faculties.TF)
+                .mapToDouble(StatementOfGrades::getGrade)
+                .average().getAsDouble() - testableAction.calculateGPAForSpecificSubjectGroupAndFaculty(
+                AcademicSubjects.ENGLISH, 2, Faculties.TF) < 0.00001);
```
[the remark this answers is not in the document]
Comment (streamdp) [reply src/test/java/com/epam/streamdp/four/actions/ActionsTest.java]: This is a reliable way to compare two double number.

File: src/main/java/com/epam/streamdp/four/actions/Actions.java [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+    public double calculateGPAForSpecificSubjectGroupAndFaculty(AcademicSubjects academicSubjects, int group, Faculties faculties)
+            throws AcademicSubjectFieldCannotBeEmptyException, TheGroupFieldMustBeSpecifiedException, TheFacultyFieldMustBeSpecifiedException {
+        statementOfGrades = statementOfGrades
+                .stream()
+                .filter(subject -> subject.getAcademicSubject() == academicSubjects)
+                .collect(Collectors.toList());
+        if (!statementOfGrades.isEmpty()) {
+            statementOfGrades = statementOfGrades
+                    .stream()
+                    .filter(subject -> subject.getStudent().getGroup() == group)
```
[the remark this answers is not in the document]
Comment (streamdp) [reply src/main/java/com/epam/streamdp/four/actions/Actions.java]: For this field I used "int", not "Integer"

File: src/test/java/com/epam/streamdp/four/actions/ActionsTest.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (yury-barsukou) [reply src/test/java/com/epam/streamdp/four/actions/ActionsTest.java]: Sorry but that is not still clear. Why do you compare two doubles? Why do you compare 1st double with 0.00001? How does comparing two doubles matches to the test name testCalculateGPAForSpecificSubjectGroupAndFaculty ? Your code/tests should not be understood only for you. Your intentions should be pretty much clear for everyone

File: src/test/java/com/epam/streamdp/four/actions/ActionsTest.java [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+import java.util.List;
+
+import static org.testng.Assert.assertEquals;
+import static org.testng.Assert.assertTrue;
+
+public class ActionsTest {
+    List<StatementOfGrades> statementOfGrades = SaveReadItemsFromJson.loadTestItemsFromFile("students_data" + File.separator + "itemForTests.json");
+    Actions testableAction = new Actions(statementOfGrades);
+
+    @Test
```
[the remark this answers is not in the document]
Comment (yury-barsukou) [reply src/test/java/com/epam/streamdp/four/actions/ActionsTest.java]: @Test(description = "Some test description")

**Pull request merged**